### PR TITLE
clean up all unhealthy replicas first for old rcs when reconcileOldRCs [LGTM]

### PR DESCRIPTION
**Pull request by mqliang**
ref https://github.com/kubernetes/kubernetes/issues/16737#issuecomment-177040867

### CODE REVIEW

**Comment by mqliang** [comment]
@bgrant0607 fix https://github.com/kubernetes/kubernetes/issues/16737 by cleanup all unhealthy replicas before scaling down healthy ones.

**Comment by bgrant0607** [comment]
Cc @kargakis @janetkuo

File: pkg/controller/deployment/deployment_controller.go [hunk 1]
```diff
@@ -630,9 +630,73 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+	totalScaledDown := cleanupCount + scaledDown
+	if expectationsCheck {
+		dc.podExpectations.ExpectDeletions(dKey, totalScaledDown)
+	}
+
+	return totalScaledDown > 0, nil
+}
+
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment) (int, error) {
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Thanks for taking a stab at this. 

The main disadvantage of this approach is that it will resize RCs to eliminate all pods that happen to be unhealthy at the moment, causing the rollout to advance unpredictably. This was the original problem we had that motivated us to use multiple RCs to perform rolling update. If we wanted this behavior, it would be simpler to just use a single RC, update its template in place, and delete the old/unhealthy pods, allowing the RC to replace them using the new template. Background is in #1353.

So, we need to limit the number of old pods we delete by resizing old RCs.

I'll describe this in terms of absolute counts for maxUnavailable and maxSurge.

Let's say the user rolled out a new revision, which deleted maxUnavailable instances of the old revision and added maxUnavailable + maxSurge instances of the new. 

However, the new revision crash loops and those pods never become ready. The user then decides to rollback.

Now the user has replicas-maxUnavailable (available) instances of the "new new" (rolled back) revision and maxUnavailable+maxSurge (unavailable) instances of the "old new" revision. We want this rollback to make progress. However, coincidentally, a node fails and causes one of the "new new" pods to fail and be recreated, but it is Pending while its image is pulled to a new node, so there are replicas-maxUnavailable-1 available instances of the "new new" revision, causing the Deployment to be over its maxUnavailable threshold. 

Intuitively, we should be able to revert maxUnavailable of the "old new" pods and delete maxSurge "old new" pods as the reverted pods become available. 

Invariants:
* There should never be more than replicas+maxSurge pods total.
* There should never be more than replicas pods of the "new new" RC.
* There shouldn't be more than maxUnavailable+maxSurge pods of the "new new" RC unavailable (though it could happen due to circumstances outside the control of the Deployment controller).
* There shouldn't be more than replicas pods available total (though there may be temporarily).

If there are fewer than maxUnavailable+maxSurge pods of the "new new" RC unavailable after having created at least maxSurge pods, then that is the budget that can be used to prune unavailable instances from old RCs.

This is not the best example because it doesn't leave pods that we don't want to delete, but hopefully this makes sense.

**Comment by mqliang** [comment]
@bgrant0607 @janetkuo Reworked(make the behaviour of rolling-update predictable). I will squash my commits if this approach looks good.

**Comment by janetkuo** [comment]
@mqliang the e2e test "deployment should support rollover" failed locally. Not sure why kubernetes-pull-build-test-e2e-gce didn't run this test? cc @ixdy

**Comment by ixdy** [comment]
@janetkuo `[Feature:]`-tagged tests (incl. deployment) are not run on PR Jenkins. @ihmccreery

File: pkg/controller/deployment/deployment_controller.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: BTW, we should document what I wrote above, in code comments and also in the design doc.

**Comment by bprashanth** [comment]
Pr jenkins as in kubernetes-jenkins? and if yes, is feature:performance different?(https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh#L549)

**Comment by bprashanth** [comment]
Ignore last comment, i misread the config file

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+		}
+		if targetRC.Spec.Replicas == readyPodCount {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		newReplicasCount := readyPodCount
+		if totalScaledDown+(targetRC.Spec.Replicas-newReplicasCount) > maxCleanupCount {
+			newReplicasCount = maxCleanupCount - totalScaledDown
+		}
```
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: I don't think the logic here is correct. Let's say we have only one old rc (available pods = 3, replicas = 5), and the maxCleanupCount = 1. After this call, we expect the rc to be scaled down to (available pods = 3, replicas = 4). 

However, based on the logic here, newReplicasCount becomes 1. Then we scale this rc down to replicas = 1, and thus available pods decreased. 

Also, can you make the logic here less complex? It's kind of hard to follow.

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxCleanupCount int) (int, error) {
 	minReadySeconds := deployment.Spec.Strategy.RollingUpdate.MinReadySeconds
-	// Check the expectations of deployment before counting available pods
-	dKey, err := controller.KeyFunc(&deployment)
+	// Safely scale down all old rcs with unhealthy replicas. ReplicationController/ReplicaSet will sort the pods in the order
+	// such that not-ready < ready, unscheduled < scheduled, and pending < running. This ensures that unhealthy replicas will
+	// been deleted first and won't increase unavailability.
+	totalScaledDown := 0
+	for _, targetRC := range oldRCs {
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Are these RCs sorted oldest to newest?

Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: No they're not.

File: pkg/controller/deployment/deployment_controller.go [hunk 4]
```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+		readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{targetRC}, minReadySeconds)
+		if err != nil {
+			return totalScaledDown, fmt.Errorf("could not find available pods: %v", err)
+		}
+		if targetRC.Spec.Replicas == readyPodCount {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		newReplicasCount := readyPodCount
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: How about:
```
scaledDown := min(maxCleanupCount - totalScaledDown, targetRC.Spec.Replicas - readyPodCount)
newReplicasCount := targetRC.Spec.Replicas - scaledDown
totalScaledDown += scaledDown
```

File: pkg/controller/deployment/deployment_controller.go [hunk 5]
```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 	}
 	// Check if we can scale down.
 	minAvailable := deployment.Spec.Replicas - maxUnavailable
+	maxScaledDown := allPodsCount - minAvailable
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: This includes both maxUnavailable and any surge pods that have been created, which is good. However, it doesn't take into account how many pods are unavailable from the newRC. 

I think we want:
```
maxScaledDown := max(allPodsCount - minAvailable - newRCPodsUnavailable, 0)
```

Concrete example:
* 10 replicas
* 2 maxUnavailable (absolute number, not percent)
* 3 maxSurge (absolute number, not percent)

Deployment is updated, newRC is created with 3 replicas, oldRC is scaled down to 8, and newRC is scaled up to 5. 

The new RC pods crashloop and never become available.

allPodsCount is 13. minAvailable is 8. newRCPodsUnavailable is 5.

A node fails and causes one of the oldRC pods to become unavailable. However, 13 - 8 - 5 = 0, so the oldRC won't be scaled down.

The user notices the crashloop and does `kubectl rollout undo` to rollback.

newRCPodsUnavailable is 1, since we rolled back to the good RC, so maxScaledDown = 13 - 8 - 1 = 4. 4 of the crashlooping pods will be scaled down. 

The total number of pods will then be 9 and the newRC can be scaled up to 10.

WDYT?

File: pkg/controller/deployment/deployment_controller.go [hunk 6]
```diff
@@ -799,28 +811,97 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+	for _, targetRC := range oldRCs {
+		if totalScaledDown >= maxCleanupCount {
+			break
+		}
+
+		if targetRC.Spec.Replicas == 0 {
+			// cannot scale down this RC.
+			continue
+		}
+		readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{targetRC}, minReadySeconds)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: RCs don't know about minReadySeconds. We should pass 0 here so that we don't cause RCs to kill available pods.

File: pkg/controller/deployment/deployment_controller.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Same example, but pushing a new pod template instead of rolling back (aka "roll over"):

The new RC created must start with 0 replicas because allPodsCount is already at 13. However, newRCPodsUnavailable would also be 0, so the 2 old RCs could be scaled down by 5 (13 - 8 - 0), which would then allow the new RC to be scaled up by 5.

Great. That's the behavior I'd want.

**Comment by mqliang** [comment]
@bgrant0607 @janetkuo comments addressed, PTAL.

File: pkg/controller/deployment/deployment_controller.go [hunk 7]
```diff
@@ -816,30 +834,102 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+		readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{targetRC}, 0)
+		if err != nil {
+			return totalScaledDown, fmt.Errorf("could not find available pods: %v", err)
+		}
+		if targetRC.Spec.Replicas == readyPodCount {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		scaledDown := int(math.Min(float64(maxCleanupCount-totalScaledDown), float64(targetRC.Spec.Replicas-readyPodCount)))
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: @thockin @lavalamp: Do you have an opinion about whether float Min/Max are appropriate for integers? I'd personally prefer creating intmin/intmax functions in pkg/util.

http://mrekucci.blogspot.com/2015/07/dont-abuse-mathmax-mathmin.html

I don't see anything in our code conventions nor in effective Go about this:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/coding-conventions.md

File: pkg/controller/deployment/deployment_controller.go [hunk 8]
```diff
@@ -816,30 +834,102 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
 	// Check if we can scale down.
 	minAvailable := deployment.Spec.Replicas - maxUnavailable
-	minReadySeconds := deployment.Spec.MinReadySeconds
-	// Check the expectations of deployment before counting available pods
-	dKey, err := controller.KeyFunc(&deployment)
-	if err != nil {
-		return false, fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	newRCPodsUnavailable := newRC.Spec.Replicas - newRCReadyPodCount
+	maxScaledDown := int(math.Max(float64(allPodsCount-minAvailable-newRCPodsUnavailable), 0))
+	if maxScaledDown <= 0 {
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Since you're checking <= 0, you could eliminate the Max with 0.

File: pkg/controller/deployment/deployment_controller.go [hunk 9]
```diff
@@ -816,30 +834,102 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 	if isPercent {
 		maxUnavailable = util.GetValueFromPercent(maxUnavailable, deployment.Spec.Replicas)
 	}
+
 	// Check if we can scale down.
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: We need more explanation here about why what we're checking is the right thing to do. Please put the explanation from my PR comments into a code comment here.

File: pkg/controller/deployment/deployment_controller.go [hunk 10]
```diff
@@ -816,30 +834,102 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
 		// Cannot scale down.
-		return false, nil
+		return 0, nil
 	}
-	totalScaleDownCount := readyPodCount - minAvailable
+
+	sort.Sort(controller.ControllersByCreationTimestamp(oldRCs))
+
 	totalScaledDown := 0
+	totalScaleDownCount := int(math.Max(float64(readyPodCount-minAvailable), float64(maxScaleDownCount)))
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Could you please explain your intent here?

File: pkg/controller/deployment/deployment_controller.go (hunk 7 above)
Comment (mqliang) [inline pkg/controller/deployment/deployment_controller.go]: +1 for adding such helpful functions.

File: pkg/controller/deployment/deployment_controller.go (hunk 10 above)
Comment (mqliang) [inline pkg/controller/deployment/deployment_controller.go]: Sorry, it's incorrect. It should be Min, instead of Max.

I'd like the total scale down count(unhealthy replicas cleaned up by `cleanupUnhealthyReplicas` +  scaled down by `scaleDownOldRCsForRollingUpdate`) doesn't exceed `maxScaledDown` (`allPodsCount - minAvailable - newRCPodsUnavailable`)

**Comment by mqliang** [comment]
@bgrant0607 @janetkuo Updated. PTAL

File: pkg/controller/deployment/deployment_controller.go [hunk 11]
```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxCleanupCount int) (int, error) {
+	sort.Sort(controller.ControllersByCreationTimestamp(oldRCs))
+	// Safely scale down all old rcs with unhealthy replicas. ReplicationController/ReplicaSet will sort the pods in the order
+	// such that not-ready < ready, unscheduled < scheduled, and pending < running. This ensures that unhealthy replicas will
+	// been deleted first and won't increase unavailability.
+	totalScaledDown := 0
+	for _, targetRC := range oldRCs {
+		if totalScaledDown >= maxCleanupCount {
```
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: `totalScaleDown` should never `> maxCleanupCount`, right?

File: pkg/controller/deployment/deployment_controller_test.go [hunk 12]
```diff
@@ -122,7 +122,7 @@ func TestDeploymentController_reconcileNewRC(t *testing.T) {
 	}
 }
 
-func TestDeploymentController_reconcileOldRCs(t *testing.T) {
+func TestDeploymentController_scaleDownOldRCsForRollingUpdate(t *testing.T) {
```
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller_test.go]: Maybe keep the test for `reconcileOldRCs`, and add tests for `scaleDownOldRCsForRollingUpdate` and `cleanupUnhealthyReplicas`?

**Comment by janetkuo** [comment]
Some questsions, LGTM otherwise. Thanks much!

File: pkg/controller/deployment/deployment_controller.go (hunk 7 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Talked to @lavalamp offline. Please create intmin/intmax functions in pkg/util/integers.

File: pkg/controller/deployment/deployment_controller.go (hunk 10 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Actually, I think we just don't want to violate minAvailable, as the code did originally. We already should have culled as many unavailable pods as possible prior to this point.

**Comment by bgrant0607** [comment]
This is close. I'd like to wrap this up this week, if possible, since we're approaching the 1.2 feature-complete deadline.

File: pkg/controller/deployment/deployment_controller.go (hunk 7 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: We could fix all the uses of float min/max in a follow-up PR.

File: pkg/controller/deployment/deployment_controller.go [hunk 13]
```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+	// * However, newRCPodsUnavailable would also be 0, so the 2 old RCs could be scaled down by 5 (13 - 8 - 0), which would then
+	// allow the new RC to be scaled up by 5.
 	minAvailable := deployment.Spec.Replicas - maxUnavailable
-	minReadySeconds := deployment.Spec.MinReadySeconds
-	// Check the expectations of deployment before counting available pods
-	dKey, err := controller.KeyFunc(&deployment)
-	if err != nil {
-		return false, fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	newRCPodsUnavailable := newRC.Spec.Replicas - newRCReadyPodCount
+	maxScaledDown := allPodsCount - minAvailable - newRCPodsUnavailable
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Can you document that maxScaledDown are the unhealthy old replicas?

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: They are now (L901)

**Comment by kargakis** [comment]
The flow looks really good

File: pkg/controller/deployment/deployment_controller.go [hunk 14]
```diff
@@ -806,9 +806,27 @@ func (dc *DeploymentController) reconcileNewRC(allRCs []*api.ReplicationControll
...
+		return false, fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	}
+	if expectationsCheck && !dc.podExpectations.SatisfiedExpectations(dKey) {
+		glog.V(4).Infof("Pod expectations not met yet before reconciling old RCs\n")
+		return false, nil
+	}
+
+	minReadySeconds := deployment.Spec.MinReadySeconds
+	allPodsCount := deploymentutil.GetReplicaCountForRCs(allRCs)
+	newRCReadyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{newRC}, minReadySeconds)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: Technically, these are "available" pods, not merely "ready" pods. Therefore, it would be useful to rename the variable to newRCAvailablePodCount.

File: pkg/controller/deployment/deployment_controller.go [hunk 15]
```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
-		glog.V(4).Infof("Pod expectations not met yet before reconciling old RCs\n")
+
+	// Clean up unhealthy replicas first, otherwise unhealthy replicas will block deployment
+	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
+	cleanupCount, err := dc.cleanupUnhealthyReplicas(oldRCs, deployment, maxScaledDown)
+	if err != nil {
 		return false, nil
 	}
+
+	totalScaledDown := cleanupCount
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: The cleaned-up pods shouldn't have reduced the number available, so cleanupCount shouldn't be relevant. The idea was to remove unproductive pods from oldRCs so that the rollout could make forward progress. But not excessively rapid forward progress, which is why we computed maxScaledDown.

At this point, we want to scale down if and only if there is a surplus of available pods (i.e., more than minAvailable).

Since scaleDownOldRCsForRollingUpdate computes minAvailable, I don't think we need to pass it another scaling limit.

File: pkg/controller/deployment/deployment_controller.go [hunk 16]
```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
 		// Cannot scale down.
-		return false, nil
+		return 0, nil
 	}
-	totalScaleDownCount := readyPodCount - minAvailable
+
+	sort.Sort(controller.ControllersByCreationTimestamp(oldRCs))
+
 	totalScaledDown := 0
+	totalScaleDownCount := int(math.Min(float64(readyPodCount-minAvailable), float64(maxScaleDownCount)))
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: maxScaleDownCount := readyPodCount - minAvailable

File: pkg/controller/deployment/deployment_controller.go [hunk 17]
```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
...
+		}
+		totalScaledDown += scaledDown
+	}
+	return totalScaledDown, nil
+}
+
+// scaleDownOldRCsForRollingUpdate scales down old rcs when deployment strategy is "RollingUpdate".
+// Need check maxUnavailable to ensure availability
+func (dc *DeploymentController) scaleDownOldRCsForRollingUpdate(allRCs []*api.ReplicationController, oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxScaleDownCount int) (int, error) {
+	maxUnavailable, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxUnavailable)
```
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: As I commented here: https://github.com/kubernetes/kubernetes/pull/20273/files#r51953975
it would be useful to factor this out into a helper function (in pkg/util/deployment).

File: pkg/controller/deployment/deployment_controller.go (hunk 11 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: It shouldn't.

This loop sums up totalScaledDown. The loop in scaleDownOldRCsForRollingUpdate subtracts from totalScaleDownCount. Please use the same approach in both places.

Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: It shouldn't.

This loop sums up totalScaledDown. The loop in scaleDownOldRCsForRollingUpdate subtracts from totalScaleDownCount. Please use the same approach in both places.

File: pkg/controller/deployment/deployment_controller.go (hunk 15 above)
Comment (mqliang) [inline pkg/controller/deployment/deployment_controller.go]: I am a little confused. Should we ensure there are at most maxScaleDown replicas been scaled down in total during one reconcileDeployment period? Or just ensure there are at most maxScaleDown unhealthy replicas been cleaned up?

File: pkg/controller/deployment/deployment_controller.go (hunk 17 above)
Comment (mqliang) [inline pkg/controller/deployment/deployment_controller.go]: I think we could address this in a follow up PR.

**Comment by mqliang** [comment]
@bgrant0607 @janetkuo unit test added. PTAL.

>Since scaleDownOldRCsForRollingUpdate computes minAvailable, I don't think we need to pass it another scaling limit.
@bgrant0607 My motivation is ensuring that there are at most maxScaleDown replicas been scaled down in total(unhealthy replicase been cleaned up + healthy replicas been scaled down) during one syncDeployment period. Do you mean we should just ensure there at most maxScaleDown unhealthy replicas been cleaned up, as for healthy replicas, do not limit it?

File: pkg/controller/deployment/deployment_controller.go (hunk 15 above)
Comment (bgrant0607) [inline pkg/controller/deployment/deployment_controller.go]: The main constraint we want to respect is that there are at least minAvailable pods available at all times. However, if the old RC pods are unavailable and don't become available, we can't make forward progress. The point of the cleanup is to enable progress when new RC pods are becoming available. maxScaleDown is only relevant to cleanup.

**Comment by mqliang** [comment]
@bgrant0607 @janetkuo Updated. PTAL.

**Comment by kargakis** [comment]
lgtm

**Comment by bgrant0607** [comment]
LGTM. Please squash now. Thanks!

**Comment by mqliang** [comment]
@bgrant0607 Squashed. PTAL

**Comment by bgrant0607** [comment]
LGTM!
